lib/fluent/plugin/exception_detector.rb

```diff
@@ -93,8 +93,17 @@ def self.supported
       rule(:ruby, /^[\t ]+.*?\.rb:\d+:in `/, :ruby)
     ].freeze
 
+    DART_RULES = [
+      rule(:start_state, /^Unhandled exception:$/, :dart_after_exception),
+      rule(:dart_after_exception,
+           /^#\d*\s*.*?\(.*?\)$|^<asynchronous suspension>$/, :dart),
+      rule(:dart_after_exception, /^.*$/, :dart_after_exception),
```

Comment:
This regex match `/^.*$/` seems quite broad. Any chance we can tighten it a bit?

Reply:
Hmmm... I did some experiment (not a Dart expert here, just playing with it).

case 1:
```
// code
void main(){
whatever line here
}

// output
Unhandled exception:
'file:///home/jdoodle.dart': error: line 2 pos 9: semicolon expected
whatever line here
        ^

#0      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:265)
#1      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

case 2:
```
// code
void main(){
throw new StateError('some message');
}

// output
Unhandled exception:
Bad state: some message
#0      main (file:///home/jdoodle.dart:2:1)
#1      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:265)
#2      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

case 3

```
// code
void main(){
throw new Error('some message');
}

// output
Unhandled exception:
NoSuchMethodError: No constructor 'Error' with matching arguments declared in class 'Error'.
Receiver: Type: class 'Error'
Tried calling: new Error("some message")
Found: new Error()
#0      NoSuchMethodError._throwNew (dart:core-patch/errors_patch.dart:196)
#1      main (file:///home/jdoodle.dart:2:16)
#2      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:265)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

Seems like it's indeed a wild match. Let's leave it then.

Reply:
Awesome,  It is wild!  The line following "Unhandled exception:" is a print of the exception thrown and in Dart any non null object can be thrown as an exception.

Reply:
On a second thought (correct me if I'm wrong), seems the format is:

`Unhandled exception:`
`one or many random lines` // Is this required?
`one or many lines starting with #\d`

If the second part is required, maybe the rules should be the following instead?
```
      rule(:start_state, /^Unhandled exception:$/, :dart_before_exception),
      rule(:dart_before_exception, /^.*$/, :dart_after_exception),
      rule(:dart_after_exception, /^.*$/, :dart_after_exception),
      rule(:dart_after_exception,
           /^#\d*\s*.*?\(.*?\)$|^<asynchronous suspension>$/, :dart),
      rule(:dart, /^#\d*\s*.*?\(.*?\)$|^<asynchronous suspension>$/, :dart)
```

Reply:
I was following the assumption that the regex rule symbols would be tested in the order they were defined. So if `rule(:dart_after_exception, /^.*$/, :dart_after_exception)` comes before any other rules for `:dart_after_exception` wouldn't this cause every following line to be considered part of the exception without stopping.  I also tried this and the tests do not pass.

Reply:
Oops. Yes, you are totally right. (Sorry for the delay. I was working on something due today on Friday and did not get a chance to look at this PR).

Reply:
Had some time to think about it, and I think @qingling128's initial intuition was correct -- these rules are too broad. The initial transition can be triggered by a line that simply reads `"Unhandled exception:"`, which isn't all that uncommon ([[1]](http://google.com/search?q="print+unhandled+exception"), [[2]](http://google.com/search?q="puts+unhandled+exception"), [[3]](http://google.com/search?q="cerr+unhandled+exception")). So imagine I'm writing a program in a language other than Dart that prints `"Unhandled exception:"` not followed by a Dart stack trace. Your rule will slurp up all of the remaining log lines under the `:dart_after_exception` rule. We don't have a way to define the expected end state, so it'll just keep adding lines until it reaches a limit, then declare victory (erm, parsing success).
We should really have tests for near misses (i.e., lines that kind of look like exception headers but aren't). The fact that we don't is not your fault, but it does mean that we have to be extra careful in selecting rules with no possibility of a runaway parse, like in this case.
@westongcooper Sorry to make you go through this, but do you want to instead add more states, one for each kind of Dart exception that could be raised (syntax error, type error, some standard error)? You'd need to add tests for each of those. I realize that it's not ideal and that this will miss some Dart exceptions, but this is a tradeoff we will have to make.

Reply:
To give an example for the concern:

```
Unhandled exception:
some random line that does not have "#"
some random line that does not have "#"
some random line that does not have "#"
... // this can go on forever
```

**Expected behavior:**
This block should not be detected as a real exception.

**Current implementation:**
The plugin will recognize this as a `Dart` exception and get stuck applying the rules, because the wildcard one always matches.

Note that we are also matching the start & end of the line (`/^Unhandled exception:$/`). So it's only an issue if someone prints exactly "Unhandled exception:" as a stand-alone line.

@westongcooper - Is this an urgent blocker for you?

BTW, alternatively we could exclude `Dart` from the default language list and require explicit turning it on with a "use with caution" notice. Not an ideal approach. But maybe a temporary workaround before we support specifying an `expected end state` rule.

Reply:
As for the potential approach to enumerate formats, an example rule to represent `SyntaxError` (4 line regex) would be:
```
rule(:dart_before_exception, /^.*\n.*\n\s*\^\n\s*$/, :dart_before_trace)
```

Reply:
Per discussion with team and internal Dark folks, we decided to propose this approach for the Dart recognizing pattern:

1. One line of `Unhandled exception:`
2. (Changing the wildcard in this part) Either `4 lines of syntax error represented by /^.*\n.*\n\s*\^\n\s*$/` or `one or many lines while every line contains ":"`
3. one of many lines with `/^#\d*\s*.*?\(.*?\)$|^<asynchronous suspension>$/`

@westongcooper - WDYT?

Reply:
Thank you for your input!  I agree with you and I can put together a more specific regex.  I'll need to take a deeper into some of the exceptions but I believe we're on the right path.  

Since any object can be thrown, not just instances of an error, the lines immediately following `Unhandled exception:` really can be anything even with many lines.   We can just hope the exceptions are subclasses of an Error and not just a `string` or `integer`.  Without an `expected end state`, there is no way of really knowing if the line `Unhandled exception:` really is the beginning of an exception.

I'll go through and make sure my regex works for typical Dart Errors.  Any other errors that do not follow a typical pattern will just not be a successful match.  I feel like it would be better to have it more stable for other languages until the support for `expected end state` arrives.

Reply:
It turns out that the plugin is 

 `...output plugin consumes a log stream of JSON objects which contain single-line log messages...`

Multi-line regex rules are not supported.

I'll have to do as @igorpeshansky mentioned and include states for a variety of exception formats.

Reply:
Yes, unfortunately you'd need to add states like `:dart_after_syntax_error_line1`, `:dart_after_syntax_error_line2`, etc, and add explicit transitions.